### PR TITLE
Web 392 forms correct validation

### PR DESCRIPTION
**Pull request by zubairali6**
These changes only apply to the contact form for now as the form action was only changed for this form to go to the controller. 

Changes to controller

- Added a check in the sendToSalesForce function to check for form errors. If there are no errors then form will be sent as normal to salesforce
- validateForm function which has various checks for validation depending on the field. 

Changes to contactform.html.twig

- Added twig conditionals to check for form errors and display span errors if there are any as per GDS standards
- Added error summary with errors and links to the appropriate fields as per GDS standards
- Changed page title depending on if there are any errors as per GDS standards
-  Inline JavaScript check to reset focus to error summary as per GDS standards

for ref - https://design-system.service.gov.uk/components/error-summary/

Screenshot 2020-09-11 at 12 23 19
Screenshot 2020-09-11 at 12 25 42

### CODE REVIEW

**Comment by Chee-Ng** [comment]
Functionality seems good, we might have to come back and decide which validation rule needs to be stronger and the exact error message too.

**Review by Chee-Ng — APPROVED** [review]

File: src/Controller/FormController.php [hunk 1]
```diff
@@ -75,19 +75,158 @@ public function sendToSalesforce(Request $request)
...
+        }
+        
+        // phone
+        if (empty($data['phone'])) {
+            $errorMessages['phoneErr']['errors'][] = 'Enter a telephone number, like 01632 960 001, 07700 900 982 or +44 0808 157 0192';
+        }
+
+        if (strlen($data['phone']) > 20) {
+            $errorMessages['phoneErr']['errors'][] = 'Telephone number must be 20 characters or fewer';
+        }
```
Comment (karishma-ccs) [inline src/Controller/FormController.php:189]: Has someone actually told us that both email and telephone number are required? 
Should we have either/or?

**Review by karishma-ccs — CHANGES_REQUESTED** [review]
Looks good overall, some minor comments below

File: src/Controller/FormController.php [hunk 2]
```diff
@@ -75,19 +75,158 @@ public function sendToSalesforce(Request $request)
...
+            'company' => $params->get('company'),
+            'jobTitle' => $params->get('00Nb0000009IXEs'),
+            'postCode' => $params->get('post-code'),
+            'moreDetail' =>  $params->get('more-detail'),
+        ];
+        
+        // check if callback checkbox has been set and add to form data
+        if ($params->get('00Nb0000009IXEg') == '1') {
+            $formData['callback'] = $params->get('00Nb0000009IXEg');
+        }
```
[the remark this answers is not in the document]
Comment (zubairali6) [reply src/Controller/FormController.php:93]: That's fine. I will change.

File: src/Controller/FormController.php [hunk 3]
```diff
@@ -75,19 +75,158 @@ public function sendToSalesforce(Request $request)
...
+                }
+            }
+        }
+    }
+
+    public function validateForm(array $data)
+    {
+        $errorMessages = [
+            'nameErr' => [
+                'errors' => [],
```
[the remark this answers is not in the document]
Comment (zubairali6) [reply src/Controller/FormController.php:133]: Yes it's for multiple errors. For instance in the email field there is an error for invalid email format and if the email is for example too long. I know this doesn't apply to the other fields but I thought I would keep it consistent and also incase we need to any more validation for these fields for whatever reason.

File: src/Controller/FormController.php [hunk 4]
```diff
@@ -75,19 +75,158 @@ public function sendToSalesforce(Request $request)
...
+            $errorMessages['companyErr']['errors'][] = 'Organisation must be 80 characters or fewer';
+        }
+
+        // job title
+        if (empty($data['jobTitle'])) {
+            $errorMessages['jobTitleErr']['errors'][] = 'Enter a job title';
+        }
+
+        if (strlen($data['jobTitle']) > 100) {
+            $errorMessages['jobTitleErr']['errors'][] = 'Job title must be 100 characters or fewer';
```
[the remark this answers is not in the document]
Comment (zubairali6) [reply src/Controller/FormController.php:206]: This length based on the HTML5 validation that was in place.

File: src/Controller/FormController.php [hunk 5]
```diff
@@ -75,19 +75,158 @@ public function sendToSalesforce(Request $request)
...
         }
+
+        // postcode
+        if (empty($data['postCode'])) {
+            $errorMessages['postCodeErr']['errors'][] = 'Enter a postcode';
+        }
```
[the remark this answers is not in the document]
Comment (zubairali6) [reply src/Controller/FormController.php:212]: We can but I assumed all fields would be required based on our converation at the time.

File: src/Controller/FormController.php (hunk 1 above)
Comment (zubairali6) [reply src/Controller/FormController.php:189]: This validation was again based on the HTML5 validation. Email will certainly be required for Pardot but telephone we can leave as optional. Although, there is an checkbox for a callback which would make the telephone a requirement I believe.

File: templates/forms/22-contact.html.twig [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {% extends 'base.html.twig' %}
 
-{% block title %}Contact Crown Commercial Services - CCS{% endblock %}
+{% block title %}{{ formErrors is defined ? 'Error: Contact Crown Commercial Services - CCS' : 'Contact Crown Commercial Services - CCS'}}{% endblock %}
```
[the remark this answers is not in the document]
Comment (zubairali6) [reply templates/forms/22-contact.html.twig:3]: This was based on the GDS standard but I can remove this to be the original title, that's fine.

File: src/Controller/FormController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (karishma-ccs) [reply src/Controller/FormController.php:133]: I think the question was more, would we want to display multiple errors at the same time on screen for a single input? It's not a great user experience.

**Pull request merged**